Add regex support

rasa opened the pull request:
Fixes #12.

I tried to write logic that converts regexs such as `(abc|defg)` to `defg`, but I gave up, as it's non-trivial. The cost/benefit is not worth it :). Instead, I just note the issue in the readme.

rasa committed:
Add regex support

rasa committed:
feat: strip out {n}s

rasa committed:
fix: ignore binary on Windows

axllent commented:
I'm sorry (and not to be rude...), but I'm really struggling to see the point of this functionality. Can you please explain the use-case for searching a regular expression?

Let me rephrase that: Can you please give me an example (or examples) of what someone may use a regular expression for?

rasa commented:
> Can you please give me an example (or examples) of what someone may use a regular expression for?

Certainly. Here are some examples:

1. `.*word.*` - find word anywhere in the key (`word.*` and `.*word` also work)
1. `^.{0,10}word` - find word anywhere in the first 10 letters of the key (how [wireguard-vanity-address](https://github.com/warner/wireguard-vanity-address) currently works)
2. `word1.*word2` - find two words, but anywhere in the key. The first word may be the hostname, the second word could be the OS, its location, whether it's a server, or just a peer, etc.
3. `(word1|word2).*(word1|word2)` - find two words, but in any order, anywhere in the key (`word1.*word1` will also match)
4. `^word[/+A-Z0-9]` - find lowercase word at beginning of key, but delimit with non-lowercase character, so `word` stands out more clearly. See also https://github.com/warner/wireguard-vanity-address/issues/22
5. `^[s5][o0][ll]ar` - find 'solar' or the visually similar 's01ar`, per https://github.com/warner/wireguard-vanity-address/issues/25
6. `^[s5][i1][z2][a4][b86][l7][e3].*` - find 'sizable' in [leet speek](https://www.gamehouse.com/blog/leet-speak-cheat-sheet/)

Since adding each letter to the search term increasing the time exponentially, we want to give the user maximum flexibility in finding the term, or terms, they are looking for.

See also https://github.com/warner/wireguard-vanity-address/issues/23

rasa commented:
I changed this to draft, as I think we should add some of the above to the readme, as I'm sure if you've questioned its usefulness, others will to.

Also, I think it's important to stop users from creating regex patterns that would never match. So, for example: `aa$` will never match, as it's missing the `=` character.

rasa committed:
fix: Unlock after loop, not function end

As defer defers to the end of the function, not the end of the block.
See https://blog.learngoprogramming.com/gotchas-of-defer-in-go-1-8d070894cb01

rasa committed:
fix: add (?i) prefix for case-insensitive regexes

axllent reviewed (commented):
I definitely would not do any `if len(..)` for either slices as this has to count both slices for every calculation, nor would I do a separate mutex for the string & regex wordmaps.

I would do something like:
```go
        // Allow only one routine at a time to avoid
	// "concurrent map iteration and map write"
	c.mapMutex.Lock()
	defer c.mapMutex.Unlock()
	for w, count := range c.WordMap {
		if count == 0 {
			continue
		}
		completed = false
		if strings.HasPrefix(matchKey, w) {
			c.WordMap[w] = count - 1
			cb(Pair{Private: k.String(), Public: pub})
		}
	}

	for w, count := range c.RegexpMap {
		if count == 0 {
			continue
		}
		completed = false
		if w.MatchString(matchKey) {
			c.RegexpMap[w] = count - 1
			cb(Pair{Private: k.String(), Public: pub})
		}
	}
```

axllent commented:
Also in `main.go`, I would bypass the estimation for regex entirely as it cannot be calculated, as well as validate the regex (rather than a `MustCompile()`)
```go
                if stripped != sword {
			regex, err := regexp.Compile(sword)
			if err != nil {
				fmt.Printf("Invalid regular expression: %s\n", sword)
				os.Exit(2)
			}
			c.RegexpMap[regex] = options.LimitResults
			fmt.Printf("Cannot calculate probability for a regular expression: %s\n", sword)
		} else {
			c.WordMap[sword] = options.LimitResults
			probability := keygen.CalculateProbability(stripped, options.CaseSensitive)
			estimate64 := int64(speed) * probability
			estimate := time.Duration(estimate64)

			fmt.Printf("Probability for \"%s\": 1 in %s (approx %s per match)\n",
				word, keygen.NumberFormat(probability), keygen.HumanizeDuration(estimate))
		}
```

rasa committed:
Skip calc for regexs, simplying loop logic

rasa committed:
go fmt

rasa committed:
Replace MustCompile with Compile

rasa committed:
Don't include added (?i) in error message

rasa committed:
Add regex section to readme, tweak regex error message

rasa committed:
Reject regexes that have no chance of finding a match

Also:
Output error messages to stderr
Only add (?i) to regex if not already there
Move regex logic to end of utils.go
Trim spaces from search term

rasa commented:
@axllent It's ready for review. Let me know your thoughts. Happy to make any changes you deem worthy.

rasa committed:
docs: Add link to regex tester, add escaping +s

axllent commented:
Thanks awesome, thanks @rasa - I did some testing and it works as I'd expect :+1:

axllent commented:
This has been merged and released in 0.0.9. I also just did a manual change to the README which resolves your other PR.

Thanks again for your hard work @rasa!
